docs/contributing/contributing-conventions.rst

```diff
@@ -0,0 +1,128 @@
+Conventions
+===========
+
+A brief overview of conventions used for the xNVMe C codebase.
+
+C: format
+---------
+
+clang-format is utilize to format the code. There are two clang-format-option
+files, one for headers (``toolbox/clang-format-h``) and one for source
+``toolbox/clang-format-c``.
+
+You can setup your editor to use the format-files, or run clang-format via the
+toolbox: ``make format-all``. This will format all files in the repository.
+
+C: style
+--------
+
+* Do not rely on compiler-specific features
+
+  - For example, gnu/gcc-only is discouraged, as it hinders portability
+  - Code should prefably translate using any of gcc/clang/icc/pgi
+  - Exceptions to this would be code in backends utilizing features which are
+    anyway only available on the platform that the backend services
+
+* Variable declaration
+
+  - At the **top** of code-blocks; before any logic, e.g.
+
+    - at the top of a function-code-block
+    - at the top of a loop-nest
+
+  - In the code-block nearest to its utilization
+
+    - in a function, the top-most scope is commonly used for
+      argument-unpacking, and the usual ``int err`` variable, along with
+      additional variable needed for the function to do its thing
+    - using variable initialization in for-loops is perfectly fine
+    - variables in nested-block-scope is fine; still at the top of that
+      block-scope, e.g. the top of a loop-nest
+
+* Variable initialization
+
+  - Do not initialize variables using functions that require error-handling
+  - Limit to argument unpacking
+  - Limit to static assignment
+
+* Global variables
+
+  - Prefix them with ``g_``
+
+C: API
+------
+
+* The API and its backend implementations aims to be re-entrant
+  - Pass pointers to "objects" e.g. ``xnvme_cmd_pass(*ctx, ...);``
+  - Pass double-pointers for "object-initialization" e.g. ``xnvme_queue(..., **queue);``
+  - On success, return 0. On error, Return negative ``errno``.
+  - Exceptions: API-calls mimicking legacy interfaces, such as the
+    ``xnvme_buf_alloc()`` which is operating similar to ``malloc()/free()``.
+
+* The API and its backend implementation cannot be assumed to be thread-safe
+
+* Be minimal with definitions in the public API
+
+  - E.g. the 'sys/queue.h' was removed from the public API as it gave several
+    head-aches when building xNVMe on multiple platforms. However, internally,
+    specifically in backend implementations, assumptions can be made on the
+    availability of certain headers and their general availability
+
+* Each function in the API must have a command-line tool exercising it
```

Comment:
It seems that many of the subsections below this are stating variants of the same thing. Maybe this can be condensed to one or two points?

Reply:
agree, tried condensing it a bit.